lib/Conversion/PipelineToHW/PipelineToHW.cpp

```diff
@@ -115,15 +123,32 @@ class PipelineLowering {
       Type dataType = regIn.getType();
       Value dataReg;
       if (this->clockGateRegs) {
-        // Clock gate based on the valid signal.
-        dataReg = builder.create<seq::CompRegClockEnabledOp>(
-            stageTerminator->getLoc(), dataType, regIn, clock, dataValid,
-            regName, reset, /*resetValue*/ Value(), /*sym_name*/ StringAttr());
+        // Use the clock gate instead of input muxing.
+        Value currClockGate = notStalledClockGate;
+        for (auto hierClockGateEnable :
+             stageTerminator.getClockGatesForReg(regIdx)) {
+          // Create clock gates for any hierarchically nested clock gates.
+          currClockGate = builder.create<seq::ClockGateOp>(
+              loc, currClockGate, hierClockGateEnable, /*test_enable=*/Value());
+        }
+        dataReg = builder.create<seq::CompRegOp>(
+            stageTerminator->getLoc(), dataType, regIn, currClockGate, regName,
+            reset, /*resetValue*/ Value(), /*sym_name*/ StringAttr());
       } else {
-        // Use input muxing.
+        // Use input muxing. The select signal is &&'ed with any clock gates
+        // that may be present.
         auto dataRegBE = bb.get(dataType);
-        auto dataRegNext = builder.create<comb::MuxOp>(
-            stageTerminator->getLoc(), dataValid, regIn, dataRegBE);
+        Value imuxSelect = stageValidAndNotStalled;
+        if (auto clockGates = stageTerminator.getClockGatesForReg(regIdx);
```

Comment:
If we don't have the clock gate regs enabled, why wouldn't we just ignore the clock gates? They're just power optimizations. If someone is using them for correctness, they're mis-using them.